[quant][graphmode][fx] Fix prepare for getitem

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57173 [quant][graphmode][fx] Fix prepare for getitem**

Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.
If getitem is called on non-Tensor inputs we'll not observe it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

jerryzh168 committed:
[quant][graphmode][fx] Fix getitem for unmatched nodes

Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Fix getitem for unmatched nodes"


Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Fix getitem for unmatched nodes"


Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Fix prepare for getitem"


Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.
If getitem is called on non-Tensor inputs we'll not observe it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -549,11 +553,18 @@ def load_arg(a: Argument) -> Argument:
+            non_tensor_input_nodes.add(node)
         if root_node is None and node.op != 'placeholder':
+            # rule 4: remove observer for getitem if it is followed by an unmatched node
+            observer_node = node.args[0]
```
should this be something like

```
maybe_observer_node = node.args[0]
if is_activation_post_process_node(maybe_observer_node, ...):
  ... do something ...
```

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -549,11 +553,18 @@ def load_arg(a: Argument) -> Argument:
             unmatched_nodes.add(node)
 
-        # rule 3: for special node, we'll just remove observer for its input
+        # rule 5: for special node, we'll just remove observer for its input
```
`getitem` is also handled here, would it make sense to combine the two?

jerryzh168 replied:
this one is removing the observer for input of getitem actually. currently the handling of getitem is a bit messy because we only want to observe the getitem nodes whose output is a Tensor, and this information is not present currently in a GraphModule, we can simplify the handling if we know the type of a Proxy.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Fix prepare for getitem"


Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.
If getitem is called on non-Tensor inputs we'll not observe it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Fix prepare for getitem"


Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.
If getitem is called on non-Tensor inputs we'll not observe it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Fix prepare for getitem"


Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.
If getitem is called on non-Tensor inputs we'll not observe it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Fix prepare for getitem"


Summary:
If getitem is followed by an unmatched node, we'll remove the observer after it.
If getitem is called on non-Tensor inputs we'll not observe it.

Test Plan:
python test/test_quantization.pyt TestQuantizeFxOps.test_getitem

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28068805](https://our.internmc.facebook.com/intern/diff/D28068805)

[ghstack-poisoned]
